File: crates/kuben-api/src/routes/registries.rs
//! Private registry logins of an environment (M4.4, ADR-030).
//!
//! A login is a managed secret of kind `registry`: encrypted revisions of a
//! username and password for one registry. Kuben resolves image tags from
//! that registry with it, and every run whose release comes from it pulls
//! with the revision it was accepted with (an immutable
//! `kubernetes.io/dockerconfigjson` Secret). The password is never returned.
//! Revisions and revocation are those of `/secrets/{name}`.

use std::collections::BTreeMap;

use axum::{
    Json,
    extract::{Path, State},
    http::StatusCode,
};
use kuben_core::{Error, perm::Perm};
use kuben_platform::secrets::RegistryLogin;
use kuben_store::repo::{SecretDeleted, SecretKind, SecretSummary};
use serde::{Deserialize, Serialize};
use utoipa::ToSchema;

use super::{
    request, scope,
    secrets::{NewRevision, store_revision},
    validate,
};
use crate::{authz::Authz, error::ApiResult, oci, state::ApiState};

/// Longest username or password accepted.
const MAX_FIELD: usize = 4096;

#[derive(Debug, Serialize, ToSchema)]
#[serde(rename_all = "camelCase")]
pub struct RegistryLoginDto {
    pub name: String,
    /// The registry as image references name it (`ghcr.io`, `docker.io`).
    pub registry: String,
    pub revision: u64,
    /// The current revision is revoked: images from the registry are not
    /// deployed until a new login is set.
    pub revoked: bool,
    pub updated_at: String,
}

#[derive(Debug, Deserialize, ToSchema)]
#[serde(deny_unknown_fields)]
pub struct PutRegistryLogin {
    /// `ghcr.io`, `docker.io`, `registry.example.com:5000`.
    pub registry: String,
    pub username: String,
    /// A password or access token; write-only.
    pub password: String,
}

impl RegistryLoginDto {
    fn of(s: SecretSummary) -> Option<Self> {
        let SecretKind::Registry(registry) = s.kind else {
            return None;
        };
        Some(Self {
            name: s.name,
            registry,
            revision: s.revision,
            revoked: s.revoked,
            updated_at: request::timestamp(s.updated_at),
        })
    }
}

/// The registry name of `given`, normalized as image references carry it.
fn registry_name(given: &str) -> Result<String, Error> {
    let given = given.trim().to_ascii_lowercase();
    let invalid = || Error::Validation(format!("`{given}` is not a registry name such as ghcr.io"));
    if given.is_empty() || given.contains('/') || given.contains('@') {
        return Err(invalid());
    }
    let parsed = oci::parse(&format!("{given}/kuben/probe")).map_err(|_| invalid())?;
    // A first part without a dot, port or `localhost` is a Docker Hub path.
    if parsed.registry != given {
        return Err(invalid());
    }
    Ok(parsed.registry)
}

fn check_login(body: &PutRegistryLogin) -> Result<(), Error> {
    for (field, value) in [("username", &body.username), ("password", &body.password)] {
        if value.is_empty() || value.len() > MAX_FIELD || value.chars().any(char::is_control) {
            return Err(Error::Validation(format!(
                "the {field} must be 1 to {MAX_FIELD} printable characters"
            )));
        }
    }
    if body.username.contains(':') {
        return Err(Error::Validation("the username cannot contain `:`".into()));
    }
    Ok(())
}

/// List the registry logins of an environment (never their passwords).
#[utoipa::path(
    get,
    path = "/projects/{project}/environments/{environment}/registries", operation_id = "listRegistryLogins",
    tag = "secrets",
    params(
        ("project" = String, Path, description = "Project name"),
        ("environment" = String, Path, description = "Environment short name"),
    ),
    responses((status = 200, body = Vec<RegistryLoginDto>))
)]
pub async fn list(
    State(state): State<ApiState>,
    authz: Authz,
    Path((project, environment)): Path<(String, String)>,
) -> ApiResult<Json<Vec<RegistryLoginDto>>> {
    let e = scope::environment(&state, &authz, &project, &environment).await?;
    let _proof = authz.require(&state, Perm::SecretRead, &e.chain())?;
    let mut tenant = state.store.tenant(e.project.org).await?;
    let logins = tenant.secrets(e.id()).await?;
    Ok(Json(
        logins.into_iter().filter_map(RegistryLoginDto::of).collect(),
    ))
}

/// Set the login of a registry: a new revision, used by the next runs.
#[utoipa::path(
    put,
    path = "/projects/{project}/environments/{environment}/registries/{name}", operation_id = "putRegistryLogin",
    tag = "secrets",
    params(
        ("project" = String, Path, description = "Project name"),
        ("environment" = String, Path, description = "Environment short name"),
        ("name" = String, Path, description = "Login name"),
    ),
    request_body = PutRegistryLogin,
    responses(
        (status = 200, body = RegistryLoginDto),
        (status = 403, body = crate::error::Problem),
        (status = 409, description = "The name or registry is taken", body = crate::error::Problem),
        (status = 422, body = crate::error::Problem),
        (status = 503, body = crate::error::Problem),
    )
)]
pub async fn put(
    State(state): State<ApiState>,
    authz: Authz,
    Path((project, environment, name)): Path<(String, String, String)>,
    Json(body): Json<PutRegistryLogin>,
) -> ApiResult<Json<RegistryLoginDto>> {
    let e = scope::environment(&state, &authz, &project, &environment).await?;
    let _proof = authz.require(&state, Perm::SecretWrite, &e.chain())?;
    validate::dns_label("login name", &name, 63)?;
    let registry = registry_name(&body.registry)?;
    check_login(&body)?;
    let values: BTreeMap<String, String> = RegistryLogin {
        username: body.username,
        password: body.password,
    }
    .values();
    let mut tenant = state.store.tenant(e.project.org).await?;
    let new = NewRevision {
        name: &name,
        kind: &SecretKind::Registry(registry),
        values: &values,
    };
    store_revision(&state, &mut tenant, &authz, &e, new).await?;
    let saved = tenant
        .secret(e.id(), &name)
        .await?
        .and_then(RegistryLoginDto::of)
        .ok_or_else(|| Error::Internal("the new registry login is missing".into()))?;
    tenant.commit().await?;
    Ok(Json(saved))
}

/// Delete a registry login. Runs accepted with it keep their revision.
#[utoipa::path(
    delete,
    path = "/projects/{project}/environments/{environment}/registries/{name}", operation_id = "deleteRegistryLogin",
    tag = "secrets",
    params(
        ("project" = String, Path, description = "Project name"),
        ("environment" = String, Path, description = "Environment short name"),
        ("name" = String, Path, description = "Login name"),
    ),
    responses(
        (status = 204, description = "Deleted"),
        (status = 404, body = crate::error::Problem),
        (status = 409, description = "An app reads it as a secret", body = crate::error::Problem),
    )
)]
pub async fn delete(
    State(state): State<ApiState>,
    authz: Authz,
    Path((project, environment, name)): Path<(String, String, String)>,
) -> ApiResult<StatusCode> {
    let e = scope::environment(&state, &authz, &project, &environment).await?;
    let _proof = authz.require(&state, Perm::SecretWrite, &e.chain())?;
    let mut tenant = state.store.tenant(e.project.org).await?;
    let not_found = || Error::NotFound(format!("registry login `{name}`"));
    let is_login = tenant
        .secret(e.id(), &name)
        .await?
        .is_some_and(|s| matches!(s.kind, SecretKind::Registry(_)));
    if !is_login {
        return Err(not_found().into());
    }
    let audit = request::audit(
        &authz,
        "secret.deleted",
        "secret",
        format!("{}/{}/{name}", e.project.slug(), e.short_name()),
    );
    match tenant.delete_secret(e.id(), &name, audit).await? {
        SecretDeleted::Done => {
            tenant.commit().await?;
            Ok(StatusCode::NO_CONTENT)
        }
        SecretDeleted::NotFound => Err(not_found().into()),
        SecretDeleted::InUse(apps) => Err(Error::Conflict(format!(
            "registry login `{name}` is read as a secret by {}",
            apps.join(", ")
        ))
        .into()),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn registries_are_named_as_image_references_name_them() {
        for (given, expected) in [
            ("ghcr.io", "ghcr.io"),
            (" GHCR.io ", "ghcr.io"),
            ("registry.example.com:5000", "registry.example.com:5000"),
            ("localhost:5000", "localhost:5000"),
            ("docker.io", "docker.io"),
        ] {
            assert_eq!(registry_name(given).expect(given), expected);
        }
        for bad in ["", "ghcr.io/acme", "acme", "ghcr.io@x", "https://ghcr.io"] {
            assert!(registry_name(bad).is_err(), "{bad:?}");
        }
    }

    #[test]
    fn logins_are_bounded() {
        let login = |username: &str, password: &str| PutRegistryLogin {
            registry: "ghcr.io".into(),
            username: username.into(),
            password: password.into(),
        };
        assert!(check_login(&login("bot", "token")).is_ok());
        assert!(check_login(&login("", "token")).is_err());
        assert!(check_login(&login("bot", "")).is_err());
        assert!(check_login(&login("a:b", "token")).is_err());
        assert!(check_login(&login("bot", "line\nbreak")).is_err());
        assert!(check_login(&login("bot", &"x".repeat(MAX_FIELD + 1))).is_err());
    }
}

File: apps/kuben/internal/httpapi/registries.go
package httpapi

// Private registry logins of an environment (routes/registries.rs, M4.4,
// ADR-030).
//
// A login is a managed secret of kind `registry`: encrypted revisions of a
// username and password for one registry. Kuben resolves image tags from
// that registry with it, and every run whose release comes from it pulls
// with the revision it was accepted with (an immutable
// `kubernetes.io/dockerconfigjson` Secret). The password is never returned.
// Revisions and revocation are those of `/secrets/{name}`.

import (
	"context"
	"strings"
	"unicode"

	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/ascii"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/kerrors"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/core/perm"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/httpapi/gen"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/integrations/oci"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/keyring"
	"github.com/Teamtem-dev/kuben/apps/kuben/internal/store"
)

// maxLoginField is the longest username or password accepted.
const maxLoginField = 4096

// registryLoginDto is RegistryLoginDto::of; false for a secret that is not
// a login.
func registryLoginDto(s store.SecretSummary) (gen.RegistryLoginDto, bool) {
	kind, ok := s.Kind.(store.SecretRegistry)
	if !ok {
		return gen.RegistryLoginDto{}, false
	}
	return gen.RegistryLoginDto{
		Name:      s.Name,
		Registry:  kind.Host,
		Revision:  revisionNumber(s.CurrentRevision),
		Revoked:   s.Revoked,
		UpdatedAt: Timestamp(s.UpdatedAt),
	}, true
}

// registryName is the registry name of given, normalized as image
// references carry it.
func registryName(given string) (string, error) {
	given = ascii.Lower(strings.TrimSpace(given))
	invalid := kerrors.New(kerrors.Validation, "`%s` is not a registry name such as ghcr.io", given)
	if given == "" || strings.ContainsAny(given, "/@") {
		return "", invalid
	}
	parsed, err := oci.Parse(given + "/kuben/probe")
	// A first part without a dot, port or `localhost` is a Docker Hub path.
	if err != nil || parsed.Registry != given {
		return "", invalid
	}
	return parsed.Registry, nil
}

// checkLogin checks the username and password of a login.
func checkLogin(body *gen.PutRegistryLogin) error {
	for _, field := range []struct{ name, value string }{
		{"username", body.Username}, {"password", body.Password},
	} {
		if field.value == "" || len(field.value) > maxLoginField || strings.ContainsFunc(field.value, unicode.IsControl) {
			return kerrors.New(kerrors.Validation, "the %s must be 1 to %d printable characters", field.name, maxLoginField)
		}
	}
	if strings.Contains(body.Username, ":") {
		return kerrors.New(kerrors.Validation, "the username cannot contain `:`")
	}
	return nil
}

// ListRegistryLogins lists the registry logins of an environment (never
// their passwords).
func (s *Server) ListRegistryLogins(ctx context.Context, params gen.ListRegistryLoginsParams) ([]gen.RegistryLoginDto, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	e, err := s.findEnvironment(ctx, a, params.Project, params.Environment)
	if err != nil {
		return nil, err
	}
	if _, err := a.Require(perm.SecretRead, e.chain()); err != nil {
		return nil, err //nolint:wrapcheck // a kerrors already
	}
	stored, err := s.storedSecrets(ctx, e)
	if err != nil {
		return nil, err
	}
	out := make([]gen.RegistryLoginDto, 0, len(stored))
	for _, sec := range stored {
		if dto, ok := registryLoginDto(sec); ok {
			out = append(out, dto)
		}
	}
	return out, nil
}

// PutRegistryLogin sets the login of a registry: a new revision, used by
// the next runs.
func (s *Server) PutRegistryLogin(
	ctx context.Context, req *gen.PutRegistryLogin, params gen.PutRegistryLoginParams,
) (gen.PutRegistryLoginRes, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	e, err := s.findEnvironment(ctx, a, params.Project, params.Environment)
	if err != nil {
		return nil, err
	}
	if _, err := a.Require(perm.SecretWrite, e.chain()); err != nil {
		return nil, err //nolint:wrapcheck // a kerrors already
	}
	if err := DNSLabel("login name", params.Name, 63); err != nil {
		return nil, err
	}
	registry, err := registryName(req.Registry)
	if err != nil {
		return nil, err
	}
	if err := checkLogin(req); err != nil {
		return nil, err
	}
	values := keyring.RegistryLogin{Username: req.Username, Password: req.Password}.Values()
	t, err := s.deps.Store.Tenant(ctx, e.project.org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // a no-op after the commit
	n := newRevision{name: params.Name, kind: store.SecretRegistry{Host: registry}, values: values}
	if err := s.storeRevision(ctx, t, a, e, n); err != nil {
		return nil, err
	}
	summary, found, err := t.Secret(ctx, e.env.ID, params.Name)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	saved, ok := registryLoginDto(summary)
	if !found || !ok {
		return nil, kerrors.New(kerrors.Internal, "the new registry login is missing")
	}
	if err := t.Commit(ctx); err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	return &saved, nil
}

// DeleteRegistryLogin deletes a registry login. Runs accepted with it keep
// their revision.
func (s *Server) DeleteRegistryLogin(ctx context.Context, params gen.DeleteRegistryLoginParams) (gen.DeleteRegistryLoginRes, error) {
	a, err := s.access(ctx)
	if err != nil {
		return nil, err
	}
	e, err := s.findEnvironment(ctx, a, params.Project, params.Environment)
	if err != nil {
		return nil, err
	}
	if _, err := a.Require(perm.SecretWrite, e.chain()); err != nil {
		return nil, err //nolint:wrapcheck // a kerrors already
	}
	name := params.Name
	t, err := s.deps.Store.Tenant(ctx, e.project.org)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	defer t.Rollback(ctx) //nolint:errcheck // a no-op after the commit
	notFound := kerrors.New(kerrors.NotFound, "registry login `%s`", name)
	current, found, err := t.Secret(ctx, e.env.ID, name)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	if _, isLogin := current.Kind.(store.SecretRegistry); !found || !isLogin {
		return nil, notFound
	}
	audit := requestAudit(a, "secret.deleted", "secret", secretReference(e, name))
	deleted, err := t.DeleteSecret(ctx, e.env.ID, name, audit)
	if err != nil {
		return nil, err //nolint:wrapcheck // a store error, answered as internal
	}
	switch d := deleted.(type) {
	case store.SecretDeletedDone:
		if err := t.Commit(ctx); err != nil {
			return nil, err //nolint:wrapcheck // a store error, answered as internal
		}
		return &gen.DeleteRegistryLoginNoContent{}, nil
	case store.SecretDeletedInUse:
		return nil, kerrors.New(kerrors.Conflict, "registry login `%s` is read as a secret by %s", name, strings.Join(d.Apps, ", "))
	case store.SecretDeletedNotFound, nil:
	}
	return nil, notFound
}
